migrate.js

```diff
@@ -0,0 +1,220 @@
+  // remove duplicates
+  const itemMap = new Map()
+  for (const item of items) {
+    itemMap.set(`${item.blockmultihash}#${item.carpath}`, item)
+  }
+
+  // map items to dynamo GetItem query keys
+  const Keys = Array.from(itemMap.values()).map(i => {
+    return {
+      blockmultihash: { S: i.blockmultihash },
+      carpath: { S: i.carpath }
+    }
+  })
+
+  const cmd = new BatchGetItemCommand({
+    RequestItems: {
+      [dst]: {
+        Keys
+      }
+    }
+  })
+
+  const res = await client.send(cmd)
+
+  /** @type {BlocksCarsPosition[]} */
+  const found = (res.Responses[dst] ?? []).map(unmarshall)
+  const foundSet = new Set()
+  for (const { blockmultihash, carpath } of found) {
+    foundSet.add(`${blockmultihash}#${carpath}`)
+  }
+
+  for (const item of itemMap.values()) {
+    const { blockmultihash, carpath } = item
+    const found = foundSet.has(`${blockmultihash}#${carpath}`)
+    if (!found) {
+      yield item
+    }
+  }
+
+  if (Object.entries(res.UnprocessedKeys ?? {}).length > 0) {
+    console.log('UnprocessedKeys', JSON.stringify(res.UnprocessedKeys))
+  }
+}
+
+/**
+ * @param {BlocksIndex} item
+ */
+export function * transformItem ({ multihash, cars }) {
+  for (const { offset, length, car } of cars) {
+    /** @type {BlocksCarsPosition} */
+    const transformed = {
+      blockmultihash: multihash,
+      // carpath: maybeUpgradeCarPath(car) ?? car,
+      carpath: car,
+      offset,
+      length
+    }
+    yield transformed
+  }
+}
+
+/**
+ * @param {string} oldPath `us-east-2/dotstorage-prod-0/raw/bafybeih5gtgdtn5fdgtessl477drjgs6silcnbbhcghe2be777syctx6le/315318734258440246/ciqcibk35zyehabw3eou3cbuharwvnhuc76uinefvu2lvoyczn4mf4y.car`
+ */
+export function maybeUpgradeCarPath (oldPath) {
+  const [, bucket, ...rest] = oldPath.split('/')
+
+  // e.g dotstorage-staging-0, dotstorage-prod-0, dotstorage-prod-1
+  if (!bucket.startsWith('dotstorage-')) {
+    // alas we have not migrated yet
+    return oldPath
+  }
+
+  const oldKey = rest.join('/')
+  const newKey = toCarKey(oldKey)
+  if (newKey) {
+    return `us-west-2/carpark-prod-0/${newKey}`
+  }
+
+  // alas we could not upgrade
+  return oldPath
+}
+
+/**
+ * Convert legacy key to car cid key where possible
+ * @param {string} key raw/bafybeieiltf3tnfdyvdutyolzhfahphgevnjsso26nulfqxtkptyefq3za/315318734258473269/ciqjxmllx5y73brw6mv3pkvd7sotfk2turkupkq7tsgygrdy2yxibri.car
+ * @returns {string | undefined} e.g bagbaieratoywxp3r7wddn4zlw6vkh7e5gkvvhjcvi6vb7henqnchrvroqdcq/bagbaieratoywxp3r7wddn4zlw6vkh7e5gkvvhjcvi6vb7henqnchrvroqdcq.car
+ */
+export function toCarKey (key) {
+  if (!key.endsWith('.car')) {
+    return undefined
+  }
+  const keyParts = key.split('/')
+  if (keyParts.at(0) === 'raw') {
+    const carName = keyParts.at(-1)
+    if (!carName) {
+      return undefined
+    }
+    const carCid = toCarCid(carName.slice(0, -4)) // trim .car suffix
+    return `${carCid}/${carCid}.car`
+  }
+  if (keyParts.at(0)?.startsWith('bag')) {
```

Comment:
Note: `bag` is not necessarily CAR e.g. https://cid.ipfs.tech/#baguqeeraobq2iwb7ccdsba3sq2dkpbff4jbzoktxdmwr4ize5jw44gkap4fq but I think will be in all buckets we have...